### ISSUE TITLE
JS UI framework-like directives

### ISSUE THREAD

**Issue by aervin** [vugu/vugu#132]
**Is your feature request related to a problem? Please describe.**
A component library would be really useful, but some core functionality in Vugu will need to be implemented first. Vue or Angular-like custom directives would be a good start, I think.

**Describe the solution you'd like**
I'm starting with a simple but common UI feature: tooltips. This is already possible to implement using Vugu's DOM Events API, but there's no simple way to reuse the implementation (that I know of) aside from copy/pasting into every component that uses a tooltip. I'd like to write something like this:
```html
<button vg-tooltip='some_expression' ...>Click me!</button>
```
where vg-tooltip is a custom directive defined by either a component library or developer. Would probably need to establish some conventions around the custom directive prefix to avoid collisions.

**Describe alternatives you've considered**
As I mentioned, the only alternative I've tried is copy/paste using the `@mouseover` and `@mouseout` DOM events.

**Additional context**
I'll share a simple Angular directive (I'm sure the Vue equivalent is known by maintainers):
```ts
import { Directive, HostListener } from '@angular/core';

@Directive({
    selector: '[tvStopPropagation]'
})
export class StopPropagationDirective {
    @HostListener('click', ['$event'])
    public onClick(event: any): void {
        event.stopPropagation();
    }
}
```
which can be used on any element like so:
```html
<button
    tvStopPropagation
    (click)="appendEmptyLDAPToForm()"
    [disabled]="!areAllLDAPFormsValid()"
>
    Submit
</button>
```

**Comment by bradleypeabody** [comment]
Thanks for bringing this up.  The use case and motivation make sense to me.  Vue has custom directives as well https://vuejs.org/v2/guide/custom-directive.html which seem to provide similar functionality.

I am however a bit concerned about the overall approach.  Here's why:

1. Practicality: In order for something like an attribute called `vg-tooltip` to work, it's meaning would have to be known at code-generation-time.  As it stands right now, the code generation actually does not know anything about any other file than the single .vugu file it is generating (except maybe a little bit of meta-data about it's package like the name - but that's really about it).  Component references for example are generated entirely based on what is written in the .vugu that is referring to the other component.  This greatly contributes to the simplicity of the overall workings of Vugu.  If `vg-tooltip` is declared in another package, there would have to be some way for the Vugu code generation to understand what it means.  Minimally it would need a package prefix, but even that I'm not sure would work just by itself.

2. Magic-ness:  With the "stop propagation" example it's pretty clear what that does, with something called "tooltip" - while I understand the desire to make it as clean as possible, the other side of it is that every one of these that is invented is some new thing or feature to learn for the developer.  My general perspective on this topic is that the things that wrap HTML/DOM should be as simple and obvious as possible, and that a bit of verbosity can be worth the long-term clarity.  Consider this alternate syntax:

```html
<button vg-events='complib.Tooltip("This is my tooltip text")'>Click me</button>
```

In this case `vg-events` would be a Vugu feature that expects an expression that returns a slice of event handlers to be added to the element.  `complib.Tooltip(...)` is a regular Go function call which returns whatever vg-events needs.   This syntax is more verbose, but in exchange, once I understand what `vg-events` does, this already tells me a great deal about what is going on, in a lot of cases it tells me everything I need to know. It also means I look up `complib.Tooltip` in regular Godoc, I don't have to go looking around for it in some other (possibly never-written) documentation.

I can foresee an issue of where some features requires multiple things - like HTML attributes and events. But if this does come up, I think it can still be worked out with a similar approach - e.g. `vg-tag=` could specify an expression which returns something that has several different methods on it to provide events, attrs and whatever else.  I would certainly want to wait until the exact use case is understood before adding this though.

I plan on starting to attempt to build out a component library soon, at which point I think more examples will crop up and the use cases will become apparent.  This will help further define the problem and allow more thought to go into the exact nature of the solution.

I also think this is a good example of where the solutions worked out for JS libraries may not translate perfectly or directly over to Go, but by considering the problem carefully we can find a solution that works well in the Go language (both in terms of literal syntax, and philosophy).

**Comment by aervin** [comment]
@bradleypeabody --

Thanks for sharing your concerns, I understand where you're coming from. The syntax you've suggested makes perfect sense to me, and I'd be happy to see that make its way into Vugu.

If you're looking for contributors to the component library, please let me know. :smiley: 

Cheers!

**Comment by bradleypeabody** [comment]
Awesome.

On that note, if you know of any CSS-only (or mostly CSS) UI libraries that could be used to get this going, that would be useful info.  My overall thought on this is that instead of trying to re-style the wheel, using an existing CSS UI library and then deciding where Vugu component wrappers make sense on top of this - I think that approach will end up better than trying to start from scratch.

Bootstrap is definitely on my list but a lot of people also like Material Design these days.  https://getmdl.io/ looks decent, I'm sure there are others.  I'm also not stuck on Material Design, and I think eventually having a number of UI styles available for Vugu would be an excellent thing.

**Comment by aervin** [comment]
My employer is a Material shop, so I've already been experimenting with CSS libraries. MDL works fine for the Material style, but I couldn't get some of the JS components to work. Admittedly, I didn't try terribly hard :sweat_smile: . A CSS-only alternative would be [Surface](https://mildrenben.github.io/surface/).

Bootstrap and Material are the big ones in my experience. Would be nice if the component library supported both to start with.

**Comment by bradleypeabody** [comment]
Yup, makes sense.  I went ahead and put this together: https://github.com/vugu/vugu/wiki/Component-Library-Notes. Hopefully that can be used to help start organizing things.  I'm not quite ready yet to dive in on this but I I do want to keep notes of the related ideas on this so the problem can be understood as well as possible by the time work starts on it.

**Comment by owenwaller** [comment]
Given that the component library never materialised, see https://github.com/vugu/vugu/issues/246#issuecomment-1780048043 do we need to look at this again post v1.0?
